### PR TITLE
Autofill: Fix issue where value would not be set correctly

### LINKED ISSUE #9831: Backspace on ComboBox is inconsistent when AllowFreeform:T
### Environment Information 
 
- __Package version(s)__: Latest (7.10.0)
- __Browser and OS versions__: Tested in Chrome and Edge on Windows

### Please provide a reproduction of the bug in a codepen:
 
https://fabricweb.z5.web.core.windows.net/oufr/7.10.0/#/examples/ComboBox
 
#### Actual behavior:
 
1. Click on first combobox example in the link above
2. Select all text with Ctrl+A
3. Hit backspace. **The entire value is cleared**
4. Select another option via keyboard or mouse
5. Tab forward into the next control
6. Tab backwards into the first combobox **The entire value is highlighted**
7. Hit backspace **The selection is cleared, but the actual value remains**
 
#### Expected behavior:
 
The outcomes of steps 3 and 7 above should be consistent

#### Priorities and help requested:
 
Are you willing to submit a PR to fix? (Yes, No)

Yes, provided I get some guidance on how to fix
 
Requested priority: (Blocking, High, Normal, Low) 
 
Normal

Products/sites affected: (if applicable)

OWA Calendar/Event compose form/Time and Time Zone fields

### PR DESCRIPTION
**Pull request by joschect**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #9831 
- [x] Include a change request file using `$ yarn change`

#### Description of changes

Fixes an issue where value would not be set correctly by autofill.

#### Focus areas to test

(optional)

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts [hunk 1]
```diff
@@ -83,8 +83,6 @@ export interface IAutofillProps extends React.InputHTMLAttributes<HTMLInputEleme
    * Handler for checking and updating the value if needed
    * in componentWillReceiveProps
    *
-   * @param defaultVisibleValue - The defaultVisibleValue that got passed
```
Comment (jspurlin) [inline packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts]: Why was this removed, is it not being used anywhere?

File: packages/office-ui-fabric-react/src/components/Autofill/Autofill.test.tsx [hunk 2]
```diff
@@ -85,4 +85,18 @@ describe('Autofill', () => {
...
+      return propsString;
+    };
+    component = mount(<Autofill componentRef={autofillRef} suggestedDisplayValue="" updateValueInWillReceiveProps={receivePropsUpdater} />);
+
+    ReactTestUtils.Simulate.input(autofill.inputElement!, mockEvent('hel'));
+    component.setProps({ suggestedDisplayValue: 'hello' });
+
+    expect(autofill.value).toBe('Updated');
+    expect(autofill.inputElement!.value).toBe('Updated');
+  });
```
Comment (jdhuntington) [inline packages/office-ui-fabric-react/src/components/Autofill/Autofill.test.tsx]: Did this test fail before the change? It looks like it should have worked...

File: packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts (hunk 1 above)
Comment (joschect) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts]: It wasn't passed into the method, so it couldn't have been used.

File: packages/office-ui-fabric-react/src/components/Autofill/Autofill.test.tsx (hunk 2 above)
Comment (joschect) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.test.tsx]: Yep, just double checked. Line 99 will fail because it's making sure that the current value of the picker is matches what's returned by updateValueInWillReceiveProps, since before the actual value of the picker was never changed.

image

File: packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts (hunk 1 above)
Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts]: Strange... is that a regression that happened at some point or was it added and just never hooked up?

Comment (joschect) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts]: No idea, but it's been that way for 2 years :/

Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts]: Hmm... maybe a follow up investigation... I wonder if there's potentially other things that might be lurking here potentially due to the fact that this wasn't being used anywhere...

Comment (ecraig12345) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts]: This whole prop is going to become a problem soon because `componentWillReceiveProps` is going away...also it's generally a bit "smelly" and makes me wonder if there's a better way to achieve whatever this was meant to achieve.

File: packages/office-ui-fabric-react/src/components/Autofill/Autofill.tsx [hunk 3]
```diff
@@ -64,16 +64,19 @@ export class Autofill extends BaseComponent<IAutofillProps, IAutofillState> impl
 
   // tslint:disable-next-line function-name
   public UNSAFE_componentWillReceiveProps(nextProps: IAutofillProps): void {
-    let newValue;
-
     if (this.props.updateValueInWillReceiveProps) {
-      newValue = this.props.updateValueInWillReceiveProps();
+      const updatedInputValue = this.props.updateValueInWillReceiveProps();
+      // Don't update if we have a null value or the value isn't changing
```
Comment (ecraig12345) [inline packages/office-ui-fabric-react/src/components/Autofill/Autofill.tsx]: What about undefined?

File: packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts (hunk 1 above)
Comment (joschect) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts]: There definitely is a better way to do this, we should refactor it so that the autofill can be a controlled component, rather than this weird callback way of handling it. All it would need to do is take in a value.

File: packages/office-ui-fabric-react/src/components/Autofill/Autofill.tsx (hunk 3 above)
Comment (joschect) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.tsx]: updateValueInWillReceiveProps is not [typed](https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts#L82-L90) as returning undefined. I can certainly add a check but to me the type is pretty explicit about what can be returned.

File: packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts (hunk 1 above)
Comment (ecraig12345) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts]: Oh gosh, *that's* what this callback was doing...??? 🤦‍♀ 🤦‍♀ 🤦‍♀

Comment (joschect) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.types.ts]: Sure seems like that's what it was doing.

**Comment by joschect** [comment]
Close and reopen to rebuild
